README.md
# IBM SoftLayer driver for Chef Provisioning

 A standalone IBM SoftLayer driver for chef-provisioning that doesn't use fog.

## Main functions

- It can provision, destroy and manipulate cloud instances on IBM SoftLayer
- Support customized system specificcation for __virtual server__, __bare metal instance__ and __bare metal server__
- Lately revision from __chef-provisioning__ v1.5


## Requirement

- A pair of IBM SoftLayer API username and key is required to use this driver
- Basic knowledge of IBM SoftLayer API, like what is virtual server, bare metal instance(hourly/monthly) and server(monthly)

## Quick Start

- Prerequisite  
  Abtain API username and key and export to environment variables or recipe (refer to __examples/cookbooks/test/recipes/virtual_server.rb__ recipe for details)

- How to run
```shell
  $ cd examples  
  $ chef-client -z -o test::virtual_server
```

## Limitation

- Assume user is responsible for correctness of specified system configuration of a cloud instance in recipe
- Due to loading meta data of IBM SoftLayer API, driver initialization might take minutes for bare metal server
- Support up to Chef 12.4

## Features/Functions to enhance

- Add 'why-run' support to help with debug or test
- Add Public/private key pair support

## Support

 Zhong Yu(Leo) Wu (leow@ca.ibm.com)   
 Emerging Technologies Team, IBM Analytics Platform (imcloud@ca.ibm.com)
 

## License

Apache v2.0

CHANGELOG.md
# IBM SoftLayer driver for chef-provisioning Changelog
## 4.0.2 (Nov/26/2015)

- Upgrade driver to link up with Chef Provisioning 1.5.0

## 3.1.10 (May/7/2015)

- Apply a patch to temporarily fix a broken function in Softlayer API

## 3.1.9 (Mar/18/2015)

- Fix regression bugs

## 3.1.1 (Mar/17/2015)

- Fix bugs in softlayer_ssh_key resource/provider

## 3.1.0 (Mar/17/2015)

- add softlayer_ssh_key provider/resource

## 3.0.1 (Feb/20/2015)

- fix typo

## 3.0.0 (Feb/20/2015)

- version alignment

## 1.0.1 (Feb/19/2015)

- Merge all outstanding features


## 1.0.0 (Feb/4/2015)

- Initial Drop



examples/cookbooks/test/recipes/virtual_server.rb
# This is a sample recipe
#
# Note: It layouts the sample recipe to show how to use SoftLayer
#       driver of Chef Provisioning to provision virtual server
#
# Authors:
#   * Zhongyu (Leo) Wu           <leow@ca.ibm.com>
#   * Emerging Technologies Team <imcloud@ca.ibm.com>
#     IBM Analytics Platform
#
# Copyright 2015, IBM Corporation
# All rights reserved
#

require 'chef'
require 'chef/provisioning'
require 'chef/provisioning/softlayer_driver/driver'

# Please fill in your username/api_key of SoftLayer to run
user_name = nil
api_key   = nil

user_name ||= ENV["API_USERNAME"]
api_key   ||= ENV["API_KEY"]

# Initialize SoftLayer driver
with_driver 'softlayer', {  :user_name => user_name,
                            :api_key   => api_key,
                            :instance_type => "virtual_server"   # instance type on SoftLayer
                                                                 # Required, String
                                                                 # One of "virtual_server", "bare_metal_instance"
                                                                 # or "bare_metal_server"
                          }

# Sample provisioner options for SoftLayer
machine_options  = {
  :billing_period =>  "hourly",            # Billing period
                                           # Required, String, "hourly"(default)
                                           # One of "hourly" or "monthly" if "virtual_server" or "bare_metal_instance"
                                           # ONLY "monthly" if "bare_metal_server"

  :datacenter  =>  "wdc01",                # Pre-defined names of data centers by SoftLayer API
                                           # Required, String, "wdc01"(default)
                                           # "wdc01", Washington DC, 1
                                           # "ams01", Amsterdam 1
                                           # "dal01", Dallas 1
                                           # "dal02", Dallas 2
                                           # "dal04", Dallas 4
                                           # "dal05", Dallas 5
                                           # "dal06", Dallas 6
                                           # "hkg02", Hong Kong 2
                                           # "lon02", London 2
                                           # "sjc01", San Jose 1
                                           # "sea01", Seattle
                                           # "sng01", Singapore 1
                                           # "tor01", Toronto 1
                                           # "mel01", melbourne 1
                                           # "par01", Paris 1
                                           # "fra02", Frankfurt 2
                                           # "mex01", Mexico 1

  :cpu => {
    :cores => 4                            # Number of cores
                                           # Required, Integer
  # :private =>                            # A flag for private cores
                                           # Optional, Boolean, False (default)
  },

  :ram =>  8,                              # Amount of RAM in GB
                                           # Required, Integer

  :os_name =>  "Ubuntu Linux 14.04 LTS Trusty Tahr - Minimal Install (64 bit)",
                                           # full description of OS name as SoftLayer Portal describe while ordering
                                           # Required, String

  :network => {                            # Network configuration
                                           # Required, Object
      :hostname => "sample-node",          # provide a host name
                                           # Required, String
      :domain   => "imdemocloud.com",      # provide a domain name
                                           # Required, String                               
      :speed    =>  1000                   # Speed of network in Mbps
                                           # Optional, Integer, 1000 (default)
                                           # One of 10, 100, 1000 if "virtual server"
  },

  :storage  => {                           # storage definition for "virtual server"
   :server_disks => [                      # disks storage
                                           # Required, Object
                                           # order of disks indicates disk index
                                           # up to 5 disks for "virtual server"
       {  
         :capacity   => 25,                # disk size in GB
         :type       => "SAN"              # disk type
                                           # One of "LOCAL" or "SAN" if virtual server
       },
       {
         :capacity   => 100,
         :type       => "SAN"
       }
    ]
  },

# :key_name                                # name for key pair created via softlayer_ssh_key
                                           # provider and resource

# :provision_timeout => 3600,              # time out for server provision in seconds
                                           # Optional, Integer
                                           # 3600  if "virtual server" 
                                           # 7200  if "bare metal instance"
                                           # 86400 if "bare metal server"

# :convergence_options => {                # convergence options defined in chef provisioning
#   :chef_version  =>                      # a customized chef client version
                                           # Optional, String
                                           # i.e. "11.16.4"

#   :ssl_verify_mode =>                    # ssl verify mode setting
                                           # optional, "verify_none" or "verify_peer"
# }

# :template_id   =>                        # SoftLayer Flex image ID (aka. GlobaleIdentifier) to provision from
                                           # Optional, String 

# :notes        =>                         # Addtional note for an instance
                                           # Optional, String
 
}

# Set up options for instance provisioning on SoftLayer
with_machine_options machine_options


# singel machine resource provision
machine "virtual-server-node" do
  tag "virtual-server-node"
  # action :destroy
  # recipe "myapp::recipe1"
  # role   "role"
end


# batch machine resource provision
#machine_batch do
#  # action :destroy
#  1.upto(2) do |i|
#    machine "biginsight-master-node-#{i}" do
#      tag "biginsight-master-node-#{i}"
#      # recipe "myapp::recipe1"
#      # role   "role"
#    end
#  end
#end


examples/cookbooks/test/recipes/baremetal_server.rb
# This is a sample recipe
#
# Note: It layouts the sample recipe to show how to use SoftLayer
#       driver of Chef Provisioning to provision bare metal server
#
#
# Authors:
#   * Zhongyu (Leo) Wu           <leow@ca.ibm.com>
#   * Emerging Technologies Team <imcloud@ca.ibm.com>
#     IBM Analytics Platform
#
# Copyright 2015, IBM Corporation
# All rights reserved
#

require 'openssl'
OpenSSL::SSL::VERIFY_PEER = OpenSSL::SSL::VERIFY_NONE

require 'chef'
require 'chef/provisioning'
require 'chef/provisioning/softlayer_driver/driver'

# Please fill in your username/api_key of SoftLayer to run
user_name = nil
api_key   = nil

user_name ||= ENV["API_USERNAME"]
api_key   ||= ENV["API_KEY"]

# Initialize SoftLayer driver
with_driver 'softlayer', {  :user_name => user_name,
                            :api_key   => api_key,
                            :instance_type => "bare_metal_server"   # One of "virtual_server", 
                                                                    #  "bare_metal_instance"
                                                                    # or "bare_metal_server"
                         }
machine_options = {
  :billing_period => "monthly",            # Billing period
                                           # Required, String, "hourly"(default)
                                           # One of "hourly" or "monthly" if "virtual_server" or "bare_metal_instance"
                                           # ONLY "monthly" if "bare_metal_server"

  :datacenter =>  "wdc01",                 # Pre-defined names of data centers by SoftLayer API
                                           # Required, String, "wdc01"(default)
                                           # "wdc01" or "Washington DC, 1"
                                           # "ams01", Amsterdam 1
                                           # "dal01", Dallas 1
                                           # "dal02", Dallas 2
                                           # "dal04", Dallas 4
                                           # "dal05", Dallas 5
                                           # "dal06", Dallas 6
                                           # "hkg02", Hong Kong 2
                                           # "lon02", London 2
                                           # "sjc01", San Jose 1
                                           # "sea01", Seattle
                                           # "sng01", Singapore 1
                                           # "tor01", Toronto 1
                                           # "mel01", melbourne 1
                                           # "par01", Paris 1
                                           # "fra02", Frankfurt 2
                                           # "mex01", Mexico 1

  :cpu => {
    :cores => 8,                          # Number of cores
                                          # Required, Integer
    :type  => "Dual Intel Xeon E5-2650 "  # full description of cpu type as SoftLayer Portal display while ordering in Web
  },

  :ram =>  16,                             # Amount of RAM in GB
                                           # Required, Integer

  :os_name =>  "Red Hat Enterprise Linux 6.x (64 bit) (per-processor licensing)",
                                           # full description of OS name as SoftLayer Portal describe while ordering
                                           # Required, String

  :network => {                            # Network configuration
                                           # Required, Object
      :hostname    => "sl-driver-test",    # provide a host name
      :domain      => "imdemocloud.com",   # provide a domain name                               
      :description => "1 Gbps Public & Private Network Uplinks"
                                           # Use text description for bare metal server to identify network spec, like:
                                           # "100 Mbps Public & Private Network Uplinks"
                                           # "100 Mbps Redundant Private Network Uplinks"
                                           # "1 Gbps Private Network Uplink"
                                           # "1 Gbps Public & Private Network Uplinks"
                                           # "1 Gbps Public & Private Network Uplinks (Unbonded)"
                                           # "1 Gbps Dual Public & Private Network Uplinks (Unbonded)""
                                           # "10 Gbps Redundant Private Network Uplinks"
                                           # "10 Gbps Public & Private Network Uplinks (Non Datacenter Restricted)"
  },

  :storage => {                           # storage configuration
      :server_disks_max => 4,             # available storage units("Up to drives")
                                          # Required if "bare_metal_server"
      :server_disks => [                  # disks configuration
                                          # Required, Array
             {
               :seq_id => 0,                     # disk index tag for storage group
               :description => "1.00 TB SATA"   # full description of disk as SoftLayer Portal describe while ordering
             },
             {
               :seq_id => 1,
               :description => "1.00 TB SATA"
             },
             {
               :seq_id => 2,
               :description => "1.00 TB SATA" 
             }
         ],
      :storage_groups => [                        # RAID configuration
                                                 # Required if bare metal server
             { :type   => "RAID 1", :hard_drives => [0,1] },
             { :type   => "JBOD",   :hard_drives => [2] }
      ]
  },

# :key_name                                # name for key pair created via softlayer_ssh_key
                                           # provider and resource

# :provision_timeout => 3600,              # time out for server provision in seconds
                                           # Optional, Integer
                                           # 3600  if "virtual server" 
                                           # 7200  if "bare metal instance"
                                           # 86400 if "bare metal server"

# :convergence_options => {                # convergence options defined in chef provisioning
#   :chef_version  =>                      # a customized chef client version
                                           # Optional, String
                                           # i.e. "11.14.0"

#   :ssl_verify_mode =>                    # ssl verify mode setting
                                           # optional, "verify_none" or "verify_peer"
# }

  # :template_id   =>                      # SoftLayer Flex image ID (aka. GlobaleIdentifier) to provision from
                                           # Optional, String 

  # :notes        =>                       # Addtional note for an instance
                                           # Optional, String
}

with_machine_options machine_options

# singel machine provision
machine "baremetal-server-node" do
  tag "baremetal-server-node"
  # action :destroy
  # recipe "myapp::recipe1"
  # role   "role"
end


# batch machines provision
#machine_batch do
#  # action :destroy
#  1.upto(2) do |i|
#    machine "sample-node-#{i}" do
#      tag "sample-node-#{i}"
#      # recipe "myapp::recipe1"
#      # role   "role"
#    end
#  end
#end



examples/cookbooks/test/recipes/baremetal_instance.rb
# This is a sample recipe for SoftLayer provisioner
#
# Note: It layouts the sample recipe to show how to invoke SoftLayer
#       driver in Chef Provisioning for bare metal instance
#
#
# Authors:
#   * Zhongyu (Leo) Wu           <leow@ca.ibm.com>
#   * Emerging Technologies Team <imcloud@ca.ibm.com>
#     IBM Analytics Platform
#
# Copyright 2015, IBM Corporation
# All rights reserved
#


require 'chef'
require 'chef/provisioning'
require 'chef/provisioning/softlayer_driver/driver'

# Please fill in your username/api_key of SoftLayer to run
user_name = nil
api_key   = nil

user_name ||= ENV["API_USERNAME"]
api_key   ||= ENV["API_KEY"]

# Initialize SoftLayer driver
with_driver 'softlayer', {  :user_name => user_name,
                            :api_key   => api_key,
                            :instance_type => "bare_metal_instance"   # instance type on SoftLayer
                                                                      # Required, String
                                                                      # One of "virtual_server", "bare_metal_instance"
                                                                      # or "bare_metal_server"
                          }

# Sample provisioner options for SoftLayer
machine_options  = {
  :billing_period =>  "hourly",            # Billing period
                                           # Required, String, "hourly"(default)
                                           # One of "hourly" or "monthly" if "virtual_server" or "bare_metal_instance"
                                           # ONLY "monthly" if "bare_metal_server"

  :datacenter  =>  "wdc01",                # Pre-defined names of data centers by SoftLayer API
                                           # Required, String, "wdc01"(default)
                                           # "wdc01", Washington DC, 1
                                           # "ams01", Amsterdam 1
                                           # "dal01", Dallas 1
                                           # "dal02", Dallas 2
                                           # "dal04", Dallas 4
                                           # "dal05", Dallas 5
                                           # "dal06", Dallas 6
                                           # "hkg02", Hong Kong 2
                                           # "lon02", London 2
                                           # "sjc01", San Jose 1
                                           # "sea01", Seattle
                                           # "sng01", Singapore 1
                                           # "tor01", Toronto 1
                                           # "mel01", melbourne 1
                                           # "par01", Paris 1
                                           # "fra02", Frankfurt 2
                                           # "mex01", Mexico 1
  :cpu => {
    :cores => 2                            # Number of cores
                                           # Required, Integer
  },

  :ram =>  2,                              # Amount of RAM in GB
                                           # Required, Integer

  :os_name => "FreeBSD 8.x (32 bit)",      # full description of OS name as SoftLayer Portal describe while ordering
                                           # Required, String
                                           # Note: only particular OS available for hourly for bare metal instance
                                           #       contact IBM SoftLayer support for detail information

  :network => {                            # Network configuration
                                           # Required, Object
      :hostname => "sample-node-test",     # provide a host name
                                           # Required, String
      :domain   => "imdemocloud.com",      # provide a domain name
                                           # Required, String                               
      :description =>  "1 Gbps Public & Private Network Uplinks"
                                           # Use text description for bare metal server to identify network spec, like:
                                           # "100 Mbps Public & Private Network Uplinks"
                                           # "100 Mbps Redundant Private Network Uplinks"
                                           # "1 Gbps Private Network Uplink"
                                           # "1 Gbps Public & Private Network Uplinks"
                                           # "1 Gbps Public & Private Network Uplinks (Unbonded)"
                                           # "1 Gbps Dual Public & Private Network Uplinks (Unbonded)""
                                           # "10 Gbps Redundant Private Network Uplinks"
                                           # "10 Gbps Public & Private Network Uplinks (Non Datacenter Restricted)"
  },

  :storage  => {                           # storage configuration for "bare metal instance"
   :server_disks => [                      # disks 
                                           # Required, Object
       {  
         :seq_id   => "0",                 # disk index
                                           # one disk ONLY if bare metal instance
         :capacity =>  500,                # disk size in GB
                                           # ONLY 500 if "bare metal instance"
         :type     => "SATA II"            # disk type
                                           # ONLY "SATA II" if "bare metal instance"
       }
    ]
  },

# :key_name                                # name for key pair created via softlayer_ssh_key
                                           # provider and resource

# :provision_timeout  => 3600,             # time out for server provision in seconds
                                           # Optional, Integer
                                           # 3600  if "virtual server" 
                                           # 7200  if "bare metal instance"
                                           # 86400 if "bare metal server"

# :convergence_options => {                # convergence options defined in chef provisioning
#   :chef_version  =>                      # a customized chef client version
                                           # Optional, String
                                           # i.e. "11.14.0"

#   :ssl_verify_mode =>                    # ssl verify mode setting
                                           # optional, "verify_none" or "verify_peer"
# }

# :template_id   =>                        # SoftLayer Flex image ID (aka. GlobaleIdentifier) to provision from
                                           # Optional, String 
                                           # i.e. "7dfe506c-e0e6-43b0-b082-96863a3926c0",

# :notes         =>                        # Addtional note for an instance
                                           # Optiona, String
 
}

# Set up options for instance provisioning on SoftLayer
with_machine_options machine_options


# singel machine resource provision
machine "baremetal-instance-node" do
  tag "baremetal-instance-node"
  # action :destroy
  # recipe "myapp::recipe1"
  # role   "role"
end


# batch machine resource provision
#machine_batch do
#  # action :destroy
#  1.upto(2) do |i|
#    machine "biginsight-master-node-#{i}" do
#      tag "biginsight-master-node-#{i}"
#      # recipe "myapp::recipe1"
#      # role   "role"
#    end
#  end
#end


lib/chef/provider/softlayer_ssh_key.rb
require 'softlayer_api'
require 'chef/provider/lwrp_base'
require 'openssl'

class Chef::Provider::SoftlayerSshKey < Chef::Provider::LWRPBase

  @@sshkeys_search_dirs = [ "#{ENV['HOME']}/.softlayer_driver/keys", 
                            "#{ENV['HOME']}/.chef/ssh", 
                            "#{ENV['HOME']}/.ssh" ]

  use_inline_resources

  def whyrun_supported?
    false
  end

  action :create do
    create_key(new_resource.name)
  end
  
  def create_key(name)
    softlayer_key = find_softlayer_key(name)
    local_key     = find_local_key(name)

    if softlayer_key
      if local_key
        raise "Public keys don't match" unless (local_key.strip == softlayer_key.strip)
      else
        raise "SSH key #{name} missing"
      end
    else
      create_local_key(name) unless local_key
      begin
        create_softlayer_key(name)
      ensure
        delete_local_key(name) unless local_key
      end
    end
  end

  protected

  def is_valid_credentail(username, api_key)
    begin
       @@client ||= SoftLayer::Client.new(username: username, api_key: api_key, timeout: 600)
       status = @@client[:Account].getAccountStatus
       return status["name"] == "Active"
    rescue => e
       raise "#{e.inspect} for user '#{username}'"
    end
    return false
  end

  def find_softlayer_key(name)
    username = new_resource.softlayer_username
    api_key  = new_resource.softlayer_api_key
    
    if username.nil? || api_key.nil?
      raise "username or api_key for SoftLayer API is blank !" 
    end

    unless is_valid_credentail(username, api_key)
      raise "Invalid username or api_key of SoftLayer API !"
    end

    @@client ||= SoftLayer::Client.new(username: username, api_key: api_key, timeout: 600)
    account = @@client[:Account]
    ssh_key = account.getSshKeys.detect { |k| k["label"] == name }
    if ssh_key
      return ssh_key["key"]
    else
      return nil
    end
  end

  # ~/.softlayer/keys
  # ~/.chef/ssh
  # ~/.ssh
  def find_local_key(name)
    
    @@sshkeys_search_dirs.each do |dir|
      public_key_path  = Dir.glob("#{dir}/#{name}.pub")
      private_key_path = Dir.glob("#{dir}/#{name}")
      next if public_key_path.empty? || private_key_path.empty?

      begin
        return ::File.read(public_key_path[0]) unless public_key_path.empty?
      rescue => e
        raise "#{e.inspect} for softlayer_ssh_key object '#{name}'"
      end
    end
    nil
  end


  # ~/.softlayer/keys/name
  # ~/.softlayer/keys/name.pub
  def create_local_key(name)    
    %x(mkdir -p #{@@sshkeys_search_dirs[0]})

    key = OpenSSL::PKey::RSA.new 2048
    type = key.ssh_type
    data = [ key.to_blob ].pack('m0')
    openssh_format = "#{type} #{data}"

    open "#{@@sshkeys_search_dirs[0]}/#{name}", 'w' do |io| io.write key.to_pem end
    open "#{@@sshkeys_search_dirs[0]}/#{name}.pub", 'w' do |io| io.write openssh_format end
  end

  def delete_local_key(name)
    @@sshkeys_search_dirs.each do |dir|
       ::File.delete("#{dir}/#{name}.pub") if ::File.exists?("#{dir}/#{name}.pub")
       ::File.delete("#{dir}/#{name}") if ::File.exists?("#{dir}/#{name}")
    end
  end

  # Design for failure
  def create_softlayer_key(name)
    public_key_path = nil
    @@sshkeys_search_dirs.each do |dir|
      public_key_path = "#{dir}/#{name}.pub"
      if ::File.exists?("#{public_key_path}")
        break
      else
        public_key_path = nil
      end
    end
    raise "No local public key found to generate SSH key in SoftLayer" if public_key_path.nil?

    public_key   = ::File.read(public_key_path)
    key_service  = @@client[:Security_Ssh_Key]
    template_key = { "key" => public_key, "label" => "#{name}" }
    key_service.createObject template_key
  end

end

lib/chef/provisioning/softlayer_driver/helpers/utils.rb
#
# Chef Provisioning driver for SoftLayer
#
# Functions: helper functions and constants
#
# Authors:
#   * Zhongyu (Leo) Wu           <leow@ca.ibm.com>
#   * Emerging Technologies Team <imcloud@ca.ibm.com>
#     IBM Analytics Platform
#
# Copyright 2015, IBM Corporation
# All rights reserved
#

class Chef
module Provisioning
    # a provisioner dedicated for IBM SoftLayer
    class SoftlayerDriver < Chef::Provisioning::Driver
      
      private

      DEFAULT_OPTIONS = {
        :provision_timeout => {
          :virtual_server      =>   3600,
          :bare_metal_instance =>   7200,
          :bare_metal_server   =>  86400
        },
        :ssh_timeout           =>    120,
        :bandwidth             =>    "500 GB Bandwidth",
        :network_speed         =>    "1000",
        :network_description   =>    "1 Gbps Public & Private Network Uplinks"
      }

      INSTANCE_IDX = ["virtual_server", "bare_metal_instance", "bare_metal_server"]
      INSTANCE_TYPES = {
        # virtual server packge id and name
        virtual_server:        { id: 46,  name: INSTANCE_IDX[0] },
        # bare metal instance package id and name
        bare_metal_instance:   { id: 50,  name: INSTANCE_IDX[1] },
        # bare metal cpu for all bare metal server packages
        bare_metal_server:     { id: 2,   name: INSTANCE_IDX[2] }
      }
      
      SERVER_STATE = {
        RUNNING:  "RUNNING",
        PASUED:   "PAUSED",
        HATLED:   "HALTED",
        ON:       "ON"
      }

      PROVISION_STATE = { 
        INIT:  "INIT",
        DONE:  "DONE" 
      }


      BILLING_TYPE = {
        hourly:   "hourly",
        monthly:  "monthly"
      }

      # categories defined in SoftLayer API
      CATEGORY_ID = {
        bms_cpu:            1,
        bms_raid:          11,
        os:                12,
        ram:                3,
        cpu:               80,
        power_supply:      35,
        bandwidth:         10,
        network:           26,
        host_ping:         20,
        remote:            46,
        primary_ip:        13,
        monitoring:        20,
        response:          22,
        notification:      21,
        vpn:               31,
        vulnerability:     32,
        disks:  [
                  "81",    # "First Hard Drive",
                  "82",    # "Second Hard Drive",
                  "92",    # "Third Hard Drive",
                  "93",    # "Fourth Hard Drive",
                  "116"    # "Fifth Hard Drive",
                ],
        bms_disks: [
                     "4",    # "First Hard Drive",
                     "5",    # "Second Hard Drive",
                     "6",    # "Third Hard Drive",
                     "7",    # "Fourth Hard Drive",
                     "36",   # "Fifth Hard Drive",
                     "37",   # "Sixth Hard Drive",
                     "38",   # "Seventh Hard Drive",
                     "39",   # "Eighth Hard Drive",
                     "40",   # "Ninth Hard Drive",
                     "41",   # "Tenth Hard Drive",
                     "42",   # "Eleventh Hard Drive",
                     "43",   # "Twelfth Hard Drive",
                     "98",   # "Thirteenth Hard Drive",
                     "99",   # "Fourteenth Hard Drive",
                     "100",  # "Fifteenth  Hard Drive",
                     "101",  # "Sixteenth  Hard Drive",
                     "102",  # "Seventeenth  Hard Drive",
                     "103",  # "Eighteenth  Hard Drive",
                     "104",  # "Nineteenth  Hard Drive",
                     "105",  # "Twentieth  Hard Drive",
                     "106",  # "Twenty-first Hard Drive",
                     "107",  # "Twenty-second Hard Drive",
                     "108",  # "Twenty-third Hard Drive",
                     "109",  # "Twenty-fourth Hard Drive",
                     "126",  # "Twenty-fifth Hard Drive",
                     "127",  # "Twenty-sixth Hard Drive",
                     "128",  # "Twenty-seventh Hard Drive",
                     "129",  # "Twenty-eighth Hard Drive",
                     "130",  # "Twenty-nineth Hard Drive",
                     "131",  # "Thirtieth  Hard Drive",
                     "132",  # "Thirty-first Hard Drive",
                     "133",  # "Thirty-second Hard Drive",
                     "134",  # "Thirty-third Hard Drive",
                     "135",  # "Thirty-fourth Hard Drive",
                     "136",  # "Thirty-fifth Hard Drive",
                     "137"   # "Thirty-sixth Hard Drive"
                    ],
        disks_max:  1000     # driver defined category
      }

      # disk attributes defined in SoftLayer API
      DISK_ATTR_TYPE_ID = {
        "RAID 0"  => 1,
        "RAID 1"  => 2,
        "RAID 5"  => 3,
        "RAID 6"  => 4,
        "RAID 10" => 5,
        "JBOD"    => 9
      }

      # cancellation reasons defined in SoftLayer API
      CANCELLATION_REASONS = {
       "unneeded"         => "No longer needed",
       "closing"          => "Business closing down",
       "cost"             => "Server / Upgrade Costs",
       "migrate_larger"   => "Migrating to larger server",
       "migrate_smaller"  => "Migrating to smaller server",
       "datacenter"       => "Migrating to a different SoftLayer datacenter",
       "performance"      => "Network performance / latency",
       "support"          => "Support response / timing",
       "sales"            => "Sales process / upgrades",
       "moving"           => "Moving to competitor"
      }
      
      def wait_until_ready(machine_spec)
          transport = nil
          count     = DEFAULT_OPTIONS[:ssh_timeout] / 60

          count.times do
            server = server_for machine_spec
            if server[:power].upcase == SERVER_STATE[:running] ||
               server[:power].upcase == SERVER_STATE[:on]
              transport = transport_for machine_spec
              return true if transport && transport.available?
            end
            sleep 60
          end

          false
      end

      # Clean up the node's IP address from the known hosts file (~/.known_hosts)
      def cleanup_host(machine_spec)
        if machine_spec.location["ip_address"]
          %x(ssh-keygen -R "#{machine_spec.location["ip_address"]}")
        end
      end

      # construct supplemental specs
      # Note: mandotry system configuration for IBM SoftLayer 
      def supplemental_specs()
        supp_specs = []
        supp_specs << { :category => :remote,        :content=> "Reboot / KVM over IP" }
        supp_specs << { :category => :primary_ip,    :content=> "1 IP Address" }
        supp_specs << { :category => :host_ping,     :content=> "Host Ping" }
        supp_specs << { :category => :response,      :content=> "Automated Notification" }
        supp_specs << { :category => :notification,  :content=> "Email and Ticket" }
        supp_specs << { :category => :vpn,           :content=> "Unlimited SSL VPN Users & 1 PPTP VPN User per account" }
        supp_specs << { :category => :vulnerability, :content=> "Nessus Vulnerability Assessment & Reporting" }
        supp_specs
      end

      # fetch datacenter price ID
      def datacenter_for(dc_name)
        datacenters = @client[:Location_Datacenter].getDatacenters
        item  = datacenters.detect { |dc| dc["longName"].include?(dc_name) || dc["name"].include?(dc_name.downcase) }
        item["id"] unless item.nil?
      end

      def service_for(instance_type)
        case instance_type
        when INSTANCE_TYPES[:virtual_server][:name]  then  "Virtual_Guest"
        else                                               "Hardware_Server"
        end
      end

      def ssh_key_for(key_name)
        ssh_key =  @client[:Account].getSshKeys.detect { |k| k["label"] == key_name }
        raise "Cannot find SSH Key #{key_name} in Softlayer account !" if ssh_key.nil?
        ssh_key["id"]
      end
    end
end
end
